### PR TITLE
Top Profiler Structured Diff

### PR DESCRIPTION
**Pull request by gme5078**
Structured Profile Diff and Structured Col Profile Diff in one PR because its easier to review.
Let me know if I mocked too much.

### CODE REVIEW

**Comment by gme5078** [comment]
Work in progress because I'm going to abstract it to the base profiler

**Comment by gme5078** [comment]
And I need type instance checking

**Comment by gme5078** [comment]
Ready for review!

File: dataprofiler/profilers/profile_builder.py [hunk 1]
```diff
@@ -1241,6 +1328,63 @@ def __add__(self, other):
...
+        :param options: options to change results of the difference
+        :type options: dict
+        :return: difference of the profiles
+        :rtype: dict
+        """
+        report = super().diff( other_profile, options)
+        report["global_stats"].update({
+                "samples_used": utils.find_diff_of_numbers(
+                    self._max_col_samples_used, 
+                    other_profile._max_col_samples_used),
```
Comment (JGSweets) [inline dataprofiler/profilers/profile_builder.py:1346]: this seems to contradict what is said in the report which uses `total_samples`

File: dataprofiler/profilers/profile_builder.py [hunk 2]
```diff
@@ -1241,6 +1328,63 @@ def __add__(self, other):
...
+
+        :param other: profile finding the difference with this one
+        :type other: StructuredProfiler
+        :param options: options to change results of the difference
+        :type options: dict
+        :return: difference of the profiles
+        :rtype: dict
+        """
+        report = super().diff( other_profile, options)
+        report["global_stats"].update({
```
Comment (JGSweets) [inline dataprofiler/profilers/profile_builder.py:1343]: what about correlations?

File: dataprofiler/profilers/profile_builder.py [hunk 3]
```diff
@@ -1241,6 +1328,66 @@ def __add__(self, other):
...
+                    self.correlation_matrix, 
+                    other_profile.correlation_matrix),
+                "profile_schema": defaultdict(list)})
+        report.update({"data_stats": []})
+
+        self_profile_schema =  defaultdict(list)
+        other_profile_schema = defaultdict(list)
+        for i in range(len(self._profile)):
+            col_name = self._profile[i].name
+            self_profile_schema[col_name].append(i)
```
Comment (JGSweets) [inline dataprofiler/profilers/profile_builder.py:1373]: how often are we doing this? should we have a util function to build this dict?

File: dataprofiler/profilers/utils.py [hunk 4]
```diff
@@ -493,6 +493,32 @@ def find_diff_of_dicts(dict1, dict2):
...
+
+    if matrix1 is not None and matrix2 is not None:
+        mat1 = np.array(matrix1, dtype=np.float)
+        mat2 = np.array(matrix2, dtype=np.float)
+        
+        if mat1.shape == mat2.shape:
+            if ((mat1 == mat2) | (np.isnan(mat1) & np.isnan(mat2))).all():
+                diff = "unchanged"
+            else:
+                diff = mat1 - mat2
```
Comment (JGSweets) [inline dataprofiler/profilers/utils.py:518]: thoughts on this vs:
```python
diff_mat = mat1 - mat2
if ((diff_mat == 0) | np.isnan(diff_mat)).all():
    diff_mat = "unchanged"
```

Comment (JGSweets) [reply dataprofiler/profilers/utils.py:518]: also allows us to set tolerances cause could do:
`np.abs(diff_mat) < tol`

File: dataprofiler/tests/profilers/test_profile_builder.py [hunk 5]
```diff
@@ -1163,6 +1163,175 @@ def test_get_and_validate_schema_mapping(self):
...
+                                    'the same profiler type.'):
+            profile.diff("ERROR")
+        
+    @mock.patch('dataprofiler.profilers.profile_builder.DataLabeler')
+    @mock.patch("dataprofiler.profilers.data_labeler_column_profile."
+                "DataLabelerColumn.update")
+    def test_diff_with_different_schema(self, *mocks):
+        input_file_path = os.path.join(
+            test_root_path, 'data', 'csv/iris.csv'
+        )
```
Comment (JGSweets) [inline dataprofiler/tests/profilers/test_profile_builder.py:1305]: i would prefer creating fake data in the test and not be externally depdendent

**Review by JGSweets — APPROVED** [review]

**Review by lettergram — APPROVED** [review]

**Pull request merged**